guidelines/commit_and_pr_text.md

```diff
@@ -0,0 +1,53 @@
+# Motivation
+Our changelogs (reviewed by users/customers) are autogenerated from PR titles and git commit text.  The better this information is, the more we help our customers.  This information also helps reviewers understand changes, patch managers understand risk/importance, and when hunting for regressions, helps us better understand what is/isn’t likely to be a cause.
+
+# PR Metadata
+
+__PR title:__ Bug NNNN: A short sentence with capitalization and no period or the Jira tag for the issue “Feature TEAM-XXX:”
+
+* Try to describe the problem the user sees, rather than the code change you are doing
+  * I.e. “Console does not work when user clicks X” instead of “Fix drawer notification integration with underlying controller”
+* Prefix bug fixes with the bug number, “Bug XXXXX:” 
+  * (capitalize bug, space, number, then colon)
+* Do not let your PR title get truncated/wrapped into the description
+* Additional prefixes like what release the PR applies to are allowed, but they will be stripped from the changelog (the changelog for 4.4 doesn’t need to say “release-4.4” 
+  * Changelog strips anything between square brackets that has no spaces - i.e. [release-4.4] is stripped but [release 4.4] would not be
+    
+
+__PR description:__  Additional context reviewers need to understand your PR
+
+* Why was this change needed?
+* What changes are being made?  
+* Reference to any prior work(other PRs) or follow up work that will be needed.
+
+[Canonical example](https://github.com/openshift/openshift-apiserver/pull/69)
+
+# Git Commit Best practices
+
+## Number of Commits
+
+Strive for a single commit per PR.
```

Comment:
If we have this guideline - Why don't we simply make squash merging the global default (with the option to override via  a label, as we already offer today)?

Reply:
i don't think we want random commit text squashed together from 15 different commits.  i'd prefer people actually think through their commit messages and commits.

Reply:
Well, the commit body you get from github on squash merge is the PR description, then a list of commits with their title and body (if any). So the `random commit text` will be at the very end of the body and can be ignored if you don't care.

Squashmerging has the advantage that it allows ppl to do small and granular commits (which can be super useful during development) without cluttering the repos history.

Using the github PR description as body rather than a single commits body has IMHO a coupe of advantages as well:
* For ppl that already write commit messages, nothing changes, because github will put the body into the pr description if its only one commit
* Ppl that are used to write proper descriptions in Github but are not used to write commit bodies do not have to change their workflow
* Reviewing the PR description is much easier than reviewing the commit body (You don't have to select the commit and yalayala/check  the pr out locally).

Reply:
I prefer squash commits too. I've configured many of my repos to be squash commit.

Reply:
On the other hand, splitting your PR into commits make them easier to review in chunks. Also, we should add text around making sure that the commits can be bisectable i.e. tests should pass at each commit.

Reply:
Squashing is good for avoiding lots of fixup commits in the history, but it is perfectly reasonable for a PR to have a well-structured series of commits (with helpful commit messages) where each commit is a separate logical change.

Reply:
For everyone commenting on this goal....It says "strive for" not "thou shalt", and is followed by a section titled "valid reasons to have multiple commits".   This is not saying we can never have multiple commits.  If you think there are additional reasons not covered by the existing list, please add suggestions to that list.  If this is a phrasing issue i'm happy to accept suggestions on a different wording that makes it clear...  the intent is "you should generally default to one commit per PR unless you fall into one of the exception categories"

I want to emphasize that in general these are guidelines not rules.   No one is going to be enforcing them strictly, we're just trying to get back to a little more discipline on our PR/git hygiene and give guidance to people who may be less experienced in what a good pr/commit looks like.